system-addon/lib/PersistentCache.jsm

```diff
@@ -0,0 +1,92 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+"use strict";
+
+const {utils: Cu} = Components;
+Cu.import("resource://gre/modules/XPCOMUtils.jsm");
+
+XPCOMUtils.defineLazyModuleGetter(this, "OS", "resource://gre/modules/osfile.jsm");
+XPCOMUtils.defineLazyGetter(this, "gTextDecoder", () => new TextDecoder());
+
+// Map of filenames to the latest data in them.
+XPCOMUtils.defineLazyGetter(this, "gInMemoryCache", () => new Map());
+// A list of cache files that has already been loaded.
+XPCOMUtils.defineLazyGetter(this, "gFilesLoaded", () => []);
+
+/**
+ * A disk based persistent cache.
+ */
+this.PersistentCache = class PersistentCache {
+
+  /**
+   * Create a cache object based on a filename.
+   *
+   * @param {string} filename Name of the file to use to persist the cache to disk.
+   */
+  constructor(filename) {
+    this.filename = filename;
+    gInMemoryCache.set(this.filename, {});
```

Comment:
This seems to duplicate an initial state that could possibly differ from that from `loadFromFile`.